04-onion-routing.md

```diff
@@ -367,6 +368,149 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Onion Messages
+
+Onion messages have an onion with an alternate `hop_payload`
+format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
+is no legacy format, thus a `bigsize` of 0 means no payload.
+
+An intermediate node expects an `enctlv` which it can decrypt using
+the `blinding` which it is handed along with the onion message.
+
+1. `tlv_stream`: `onionmsg_payload`
+2. types:
+    1. type: 2 (`reply_path`)
+    2. data:
+        * [`point`:`first_node_id`]
+        * [`point`:`blinding`]
+        * [`...*onionmsg_path`:`path`]
+    1. type: 10 (`enctlv`)
+    2. data:
+        * [`...*byte`:`enctlv`]
+
+1. `tlv_stream`: `encmsg_tlvs`
+2. types:
+    1. type: 1 (`padding`)
+    2. data:
+        * [`...*byte`:`pad`]
+    1. type: 4 (`next_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+    1. type: 12 (`next_blinding`)
+    2. data:
+        * [`point`:`blinding`]
+    1. type: 14 (`self_id`)
+    2. data:
+        * [`...*byte`:`data`]
```

Comment:
Isn't the `self_id` intended for the final recipient? I thought that `enctlv` was only for intermediate nodes and `reply_path` only for the final recipient. If that's the case, then `self_id` should be in the `onionmsg_payload` next to `reply_path` and not nested inside a `enctlv` that is not needed for the final node.

Reply:
enctlv is optional for the final recipient, but in practice you need it, to contain *some* identifying information (proposed here as `self_id`) as soon as you start using onion messages.  Otherwise you need to remember the blinding factor for every single path you've ever sent out, to avoid responding inappropriately.

Reply:
I understood why you want the `self_id`, I just thought it would be simpler to put it directly in the `onionmsg_payload`, not in the `enctlv`. I've now realized that it needs to be in the `enctlv` from the `reply_path` so that the `self_id` can't be modified by the person who replies.